### PR TITLE
[SPARK-32124][CORE][FOLLOW-UP] Use the invalid value Int.MinValue to fill the map index when the event logs from the old Spark version

### PR DESCRIPTION
**Pull request by xuanyuanking**
### What changes were proposed in this pull request?
Use the invalid value Int.MinValue to fill the map index when the event logs from the old Spark version.

### Why are the changes needed?
Follow up PR for #28941.

### Does this PR introduce _any_ user-facing change?
When we use the Spark version 3.0 history server reading the event log written by the old Spark version, we use the invalid value -2 to fill the map index.

### How was this patch tested?
Existing UT.

### CODE REVIEW

**Comment by xuanyuanking** [comment]
cc @warrenzhu25 @HeartSaVioR @dongjoon-hyun

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala
Comment (warrenzhu25) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:1086]: Any special reasons using `-2` instead of `-1`? BTW, I'm unsure whether this filed will be displayed in UI, if so, we better show `Unknown` for invalid value in UI.

Comment (HeartSaVioR) [reply core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:1086]: Yeah once we provide it with magic number to mark INVALID/UNKNOWN, there should be relevant code change to handle INVALID/UNKNOWN value. Looks like the code change doesn't address it, which makes me unsure how it will be represented on the UI.

Comment (gengliangwang) [reply core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:1086]: @HeartSaVioR I think it is represented as error string in `FetchFailed.toErrorString`. So we just need to make sure the default unknown value is consistent in the code here and `FetchFailed.toErrorString`. Personally I prefer `Int.MinValue` since `-1` can't be used here.

Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:1086]: Thanks Gengliang for confirming!
```
Any special reasons using -2 instead of -1?
```
mapIndex = -1 has been used in `MetadataFetchFailedException`.

Yep, I think `Int.MinValue` is a better choice.

Comment (HeartSaVioR) [reply core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: Thanks for the info! If the value is only leveraged via `FetchFailed.toErrorString`, what we may want to do (and proposed in previous comment) is projecting the magic number to the specific string so that it can be determined easily. +1 to `Int.MinValue`.

Comment (HeartSaVioR) [reply core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: And I'm seeing it's done. Thanks for the quick address!

Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: Yep, already addressed in 9dfa5c0

**Comment by dongjoon-hyun** [comment]
Hi, @HeartSaVioR . Could you finalize this as an Apache Spark committer? :)

**Comment by gengliangwang** [comment]
@HeartSaVioR Congratulations!

**Review by HeartSaVioR — APPROVED** [review]
LGTM pending Jenkins

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
Btw thanks @dongjoon-hyun and @gengliangwang ! I thought it's probably better to take committer's work after official announcement, but this looks to be simple and good example to try out before dealing with larger PR. Thanks again for the guidance!

**Comment by dongjoon-hyun** [comment]
Ya. That's also reasonable. Usually, the annoucement gets some delays because the invitations go out to multiple committers at the same season. You don't need to be blocked. ;) You've been waiting so long. Please do your what you want.

**Comment by HeartSaVioR** [comment]
Jenkins doesn't seem to pick this up by request ;( Mentioned the behavior in the dev@ list.

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
retest this please

**Comment by xuanyuanking** [comment]
Wow, Congratulations @HeartSaVioR! Well deserved!

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by xuanyuanking** [comment]
retest this please

**Comment by HeartSaVioR** [comment]
retest this please

**Comment by HeartSaVioR** [comment]
retest this please

**Comment by HeartSaVioR** [comment]
Thanks! Merged to master and branch-3.0.

**Comment by HyukjinKwon** [comment]
The very first commit merged by @HeartSaVioR!

**Pull request closed without merging**
